### PR TITLE
cubehelix established

### PR DESCRIPTION
**Pull request by lobingera**
Following #78 i tried to build a cubehelix function (works with e.g. Winston.colormap). Could use some more documentation and while closely looking at it: there is a possiblility to integrate with the colormap("name",N) call.

### CODE REVIEW

**Comment by lobingera** [comment]
Is there anything i need to improve here? Any comment would be nice.

**Comment by kmsquire** [comment]
LGTM.  I don't usually do much with the color package, but this seems to belong here, and the original author is clearly interested in broad scale implementation.  If no one objects, I'll merge this tomorrow.

**Comment by kmsquire** [comment]
(or gets to it first)

File: src/cubehelix.jl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import Color: RGB
+
+# parameters: N number of entries in result colormap (btw: Array(RGB{Float64},N))
+#  rots: rotations -> added term
+#  start: startpoint in H -> 0.0 -> Red
+#  hue: saturation of hue -> 0.0 -> grayscale only, >1.0 -> clipping in some areas
+#  gamma: gamma factor 
+
+function cubehelix(N::Int,rots::Real,start::Real,hue::Real,gamma::Real)
```
Comment (stevengj) [inline src/cubehelix.jl]: probably only `Integer`, not `Int`, is fine here

Comment (stevengj) [inline src/cubehelix.jl]: Equivalent to your defaults, you can just do:
```
function cubehelix(N::Integer,rots::Real=0.5,start::Real=0,hue::Real=1,gamma::Real=1)
```
However, I would tend to use keywords in this situation: positional arguments get confusing when there are so many arguments, and the function-call overhead (which is worse with keywords) is not an issue here.  i.e.
```
function cubehelix(N::Integer; rots::Real=0.5,start::Real=0,hue::Real=1,gamma::Real=1)
```
(note the semicolon).   Is there a more descriptive variable name than `rots`?

Comment (stevengj) [inline src/cubehelix.jl]: It looks like there should also be some argument-checking (e.g. do the arguments have to be in [0,1]?); you can throw an `ArgumentError` if an argument is out of range.

File: src/cubehelix.jl [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    colormap = Array(RGB{Float64},N)
+    fract_all = linspace(0,1,N)
+    coeff_cos = [+0.14861, -0.29227, +1.97294]
+    coeff_sin = [+1.78277, -0.90649, 0.0]
+    for fe in enumerate(fract_all)
```
Comment (stevengj) [inline src/cubehelix.jl]: `for (i,fract) in enumerate(fract_all)` would be cleaner.

File: src/cubehelix.jl [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        fract = fe[2]
+        i = fe[1]
+        angle = 2*pi*(((start/3.0)+1.0+rots)*fract)
+        fract = fract ^ gamma
+        amp = hue*fract*(1-fract)/2.0
+        c = zeros(3)
```
Comment (stevengj) [inline src/cubehelix.jl]: If performance matters (probably it doesn't), you should allocate `c` outside this loop and just set `c[1]=c[2]=c[3]=0` here.

File: src/cubehelix.jl [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# parameters: N number of entries in result colormap (btw: Array(RGB{Float64},N))
+#  rots: rotations -> added term
+#  start: startpoint in H -> 0.0 -> Red
+#  hue: saturation of hue -> 0.0 -> grayscale only, >1.0 -> clipping in some areas
+#  gamma: gamma factor 
+
+function cubehelix(N::Int,rots::Real,start::Real,hue::Real,gamma::Real)
+    colormap = Array(RGB{Float64},N)
+    fract_all = linspace(0,1,N)
+    coeff_cos = [+0.14861, -0.29227, +1.97294]
```
Comment (stevengj) [inline src/cubehelix.jl]: Should these numbers have more digits?

**Comment by stevengj** [comment]
Also, there is no copyright information on the original Fortran source, so distributing derived works is technically illegal under US copyright law.  (The *default* in copyright law is *no* permission to redistribute.)  A lot of scientists don't understand this, and think that if they just post their code on the web then it is free for anyone to use without restriction.

Probably, this was not the original author's intention, so the best thing to do would be to email him and get permission to use/distribute/modify/translate his code under the terms of the [MIT/expat license](http://opensource.org/licenses/MIT).

File: src/cubehelix.jl [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        colormap[i] = RGB{Float64}(c[1],c[2],c[3])
+    end
+    colormap
+end
+
+# simpler calls with defaults
+cubehelix(N::Int,rots::Real,start::Real,hue::Real) = cubehelix(N::Int,rots::Real,start::Real,hue::Real,1.0)
+cubehelix(N::Int,rots::Real,start::Real) = cubehelix(N::Int,rots::Real,start::Real,1.0,1.0)
+cubehelix(N::Int,rots::Real) = cubehelix(N::Int,rots::Real,0.0,1.0,1.0)
+cubehelix(N::Int) = cubehelix(N::Int,0.5,0.0,1.0,1.0)
```
Comment (stevengj) [inline src/cubehelix.jl]: the author seems to suggest defaulting rotations to -1.5 on his web page...why 0.5?

File: src/cubehelix.jl (hunk 1 above)
Comment (lobingera) [inline src/cubehelix.jl]: I agree with all of the above. The only limited valid range of input arguments seems to be hue and gamma.

File: src/cubehelix.jl (hunk 2 above)
Comment (lobingera) [inline src/cubehelix.jl]: Agreed, actually with a python background i should have seen that (i wasn't aware that a valid julia concept)

File: src/cubehelix.jl (hunk 3 above)
Comment (lobingera) [inline src/cubehelix.jl]: Agreed.

File: src/cubehelix.jl (hunk 4 above)
Comment (lobingera) [inline src/cubehelix.jl]: Good question. Actually this seemed to be angles in rad, i'll try to decode from the publication. However, this seems to be fortran REAL, so single precision, so only 6 valid digits in decimal...

File: src/cubehelix.jl (hunk 5 above)
Comment (lobingera) [inline src/cubehelix.jl]: A send an email to Dave Green a few minutes ago, let's see his comment. And you are right, the web page suggests 'his' default. I think i have seen 0.5 in one of the other references. I'm fine with -1.5 also.

Comment (lobingera) [inline src/cubehelix.jl]: Sending an email to Cambridge seems to be include some problems:
```
------ This is a copy of the message, including all the headers. ------

Return-path: <lobingera@gmail.com>
Received: from ppsw-41.csi.cam.ac.uk ([131.111.8.141])
        by mail.mrao.cam.ac.uk with esmtps (TLSv1:AES256-SHA:256)
        (Exim 4.63)
        (envelope-from <lobingera@gmail.com>)
        id 1ZAxNX-000227-CT
        for dag@mrao.cam.ac.uk; Fri, 03 Jul 2015 10:35:27 +0100
X-Cam-SpamDetails: score -0.7 from SpamAssassin-3.3.2-1687995
 * -0.7 RCVD_IN_DNSWL_LOW RBL: Sender listed at http://www.dnswl.org/, low
 *      trust
 *      [209.85.215.45 listed in list.dnswl.dnsbl.ja.net]
 *  0.0 FREEMAIL_FROM Sender email is commonly abused enduser mail provider
 *       (lobingera[at]gmail.com)
```

Shall i inform the google about it?

Comment (stevengj) [inline src/cubehelix.jl]: Probably google can't do much about it; it's a question for the mail admins at `mail.mrao.cam.ac.uk`.  Did they actually reject your message or just warn you?  I could try sending from an mit.edu domain.

Comment (lobingera) [inline src/cubehelix.jl]: afaics the message(s) are rejected. Initially i even used html unknowingly which give a worse score...

What do you mean with trying to send from a mit.edu domain?

Comment (cormullion) [inline src/cubehelix.jl]: @lobingera  the mit.edu domain is white-listed at dnswl.org and emails from here should get past spamassassin; however, gmail.com is considered to be the source of much spam :) so your email won't get through...

Comment (lobingera) [inline src/cubehelix.jl]: Actually i understood, that sending emails from other domains should get high enough ranking. But somehow i'm not even remotely associated with MIT, is there a possibility to get email service as outsider?

Comment (stevengj) [inline src/cubehelix.jl]: I meant that I would email the author for you; I have an mit.edu account.

Comment (lobingera) [inline src/cubehelix.jl]: Let's hope mit.edu accepts from the google.

**Comment by lobingera** [comment]
News: My original email was delivered (David Green found it in the spam folder). He replied: "I believe the paper I published on this, which includes the fortran
implementation, is in the public domain. I'll check with the journal."

So let's wait for his next email.

**Comment by stevengj** [comment]
It's almost certainly *not* in the public domain (== *not copyrighted*), because everything is copyrighted by default (at least in the US) and publishers rarely relinquish this.   One key question is whether he signed a copyright assignment form when he published the paper—he could have relinquished his rights to the code.  (Authors often do this without fully understanding what they are doing, in my experience.)

Can you just request that he attach an [explicit MIT/expat license](http://opensource.org/licenses/MIT) statement to his code, if it is his intention that the code be freely usable/redistributable/modifiable without restriction?  Please make sure he understands that the *default* under US (and other international) copyright laws is *no permission*, so an *explicit* license statement is needed to ensure that the code is free of legal encumbrance for all time.  And it is *much* better to use a standard, time-tested license statement than it is to write your own informal language (e.g. "this code is public domain") whose legal meaning is often (unintentionally) unclear or (worse) has unintended consequences.

**Comment by stevengj** [comment]
And please cc me (stevenj at math dot mit dot edu) on your email to him.

**Comment by lobingera** [comment]
@stevengj, @kmsquire 
We're now somehow waiting 4 weeks for a answer about this copyright issue Steven brought up. Would it make some sense to ask for a volunteer who is not aware of neigther my code nor the cited f77 to write a 4 liner that fills the palette?

**Comment by stevengj** [comment]
@lobingera, yes, it seems time for a clean-room implementation based on the description in the paper only.   Maybe ask for volunteers on julia-users?  As you say, the implementation should be trivial.

**Comment by lobingera** [comment]
News: David Green's website now has a license attached to the code which afaiu put's the code into public domain.

**Comment by lobingera** [comment]
Will be moved to https://github.com/JuliaGraphics/Colors.jl

**Pull request closed without merging**